### PR TITLE
Minor selectHost optimization in scheduler

### PR DESCRIPTION
**Pull request by olegshaldybin**
Instead of sorting hosts by score, find max score and choose one of
the hosts with max score directly. Saves a little time on sorting and
avoids extra copying of HostPriorityList entries.

Test had to be updated as one of the test cases relied on having a
stable order from pre-sorting.

### CODE REVIEW

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 1]
```diff
@@ -102,20 +101,34 @@ func (g *genericScheduler) Schedule(pod *api.Pod, nodeLister algorithm.NodeListe
...
-	sort.Sort(sort.Reverse(priorityList))
 
-	hosts := getBestHosts(priorityList)
+	maxScore := priorityList[0].Score
+	// idx contains indices of elements with score == maxScore.
+	idx := []int{}
+
+	for _, entry := range priorityList {
+		if entry.Score > maxScore {
+			maxScore = entry.Score
```
Comment (lavalamp) [inline plugin/pkg/scheduler/generic_scheduler.go]: You can do it in one pass if you do idx = idx[:0] here...

**Comment by lavalamp** [comment]
I don't expect this to make much difference in scheduler run time, honestly I'm not sure it's worth it. There's O(N*M) sections to the scheduler (pods x nodes); this doesn't change the overall asymptotic running time. And there's only like 1000 nodes, so I'm not sure this is a big factor.

...I'm not necessarily opposed to this, either, since it doesn't make it more complicated. A benchmark of some sort might be nice. There's one in the integration test suite that you could maybe just make before-and-after runs with?

**Comment by olegshaldybin** [comment]
Thanks for taking a look! I also don't anticipate it making a lot of difference, was mostly poking around scheduler implementation and tweaking things to get a better understanding. I'm out of town this week, will benchmark when I'm back.

File: plugin/pkg/scheduler/generic_scheduler.go (hunk 1 above)
Comment (krousey) [inline plugin/pkg/scheduler/generic_scheduler.go]: Agreed, this should be done in one pass.

**Comment by krousey** [comment]
@k8s-bot ok to test

**Comment by krousey** [comment]
@k8s-bot ok to test

**Comment by olegshaldybin** [comment]
@krousey @lavalamp Good point, I changed selectHost to be a single-pass. As expected, it doesn't change the overall scheduling latency by much, as it's still dominated by predicates/priority functions. It does however decrease selectHost run time from about 400μs to 12μs in my tests with 1000 nodes. Leaving it up to you to decide whether this kind of optimization makes sense.

One other thing I've noticed is that the scheduler benchmark has a pretty high variance, up to 20% in my tests, so it's hard to detect the actual impact of small changes. I will try to take a closer look at the source of that variance.

**Comment by krousey** [comment]
@k8s-bot e2e test this #20631 #20403

**Comment by wojtek-t** [comment]
@olegshaldybin 

/workspace/kubernetes/docs/api-reference is out of date. Please run hack/update-api-reference-docs.sh

**Comment by olegshaldybin** [comment]
@wojtek-t I rebased this branch on master, I assume this should pick up the updated docs?

**Pull request merged**